midend/flattenLogMsg.cpp

```diff
@@ -0,0 +1,198 @@
+#include "flattenLogMsg.h"
+
+namespace P4 {
+
+const IR::Type_StructLike* FindTypeInLogMsgForReplace::hasStructInParameter(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* path = methodCallStatement->methodCall->method->to<IR::PathExpression>()) {
+        if (path->path->name.name == "log_msg") {
+            auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+            auto* param2 = methodCallStatement->methodCall->arguments->at(1)->expression;
+            if (!param1->is<IR::StringLiteral>() || !param2->is<IR::StructExpression>()) {
+                return nullptr;
+            }
+            auto* type =
+                typeMap->getTypeType(methodCallStatement->methodCall->typeArguments->at(0), true);
+            if (auto* typeStruct = type->to<IR::Type_StructLike>()) {
+                for (auto field : typeStruct->fields) {
+                    if (field->type->is<IR::Type_StructLike>()) {
+                        return typeStruct;
+                    }
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+const IR::MethodCallStatement* FindTypeInLogMsgForReplace::prepareLogMsgStatement(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* newMethod = getReplacementMethodCall(methodCallStatement->id)) {
+        return newMethod;
+    }
+    // Create new statement
+    auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+    std::string strParam1 = param1->to<IR::StringLiteral>()->value.c_str();
+    auto* argType =
+        typeMap->getTypeType(methodCallStatement->methodCall->typeArguments->at(0), true);
+    auto* structType = argType->to<IR::Type_StructLike>();
+    index = 0;
+    auto exprVector =
+        unfoldStruct(methodCallStatement->methodCall->arguments->at(1)->expression, strParam1);
+    structType = generateNewStructType(structType, exprVector.first);
+    auto* newMethodCall = methodCallStatement->methodCall->clone();
+    auto* newTypeArguments = newMethodCall->typeArguments->clone();
+    newTypeArguments->clear();
+    newTypeArguments->push_back(structType);
+    newMethodCall->typeArguments = newTypeArguments;
+    auto* newArguments = newMethodCall->arguments->clone();
+    auto* newArgument = newMethodCall->arguments->at(0)->clone();
+    auto* newString = newArgument->expression->to<IR::StringLiteral>()->clone();
+    newString->value = exprVector.second;
+    newArgument->expression = newString;
+    newArguments->at(0) = newArgument;
+    newArgument = newMethodCall->arguments->at(1)->clone();
+    auto* oldType = newArgument->expression->to<IR::StructExpression>()->structType;
+    auto newStructExpression =new IR::StructExpression(oldType->srcInfo, structType,
+        structType->getP4Type()->template to<IR::Type_Name>(), exprVector.first);
+    newStructExpression->components = exprVector.first;
+    typeMap->setType(structType, structType);
```

Comment:
why do you need to set these types? you will clear the type map immediately after this.